sycl/source/detail/non_cached_kernel_lock.cpp

```diff
@@ -0,0 +1,19 @@
+#include "non_cached_kernel_lock.hpp"
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+NonCachedKernelLock::~NonCachedKernelLock() {}
+
+Locked<RT::PiKernel> NonCachedKernelLock::lockKernel(RT::PiKernel &K) {
+  std::unique_lock<std::mutex> MapLock{MapMtx};
+  std::mutex &Mtx = Map[K];
+  MapLock.unlock();
```

Comment:
Thinking how we could prettify it, maybe something like this?:
```suggestion
  std::mutex &Mtx = [&]() -> std::mutex & { 
       std::lock_guard<std::mutex> MapLock{MapMtx};
       return Map[K];
   }();
```
PS. Didn't run the snippet thru compiler.

Reply:
This code snippet looks like being working to me. At least the idea behind it. Although, there's an overhead of creating a lambda object and calling it.

Reply:
I would expect that this code is always optimized by the compiler: https://godbolt.org/z/c95WbjnWq .